### PR TITLE
feat: Add --dump-ast option to dump the json representation.

### PR DESCRIPTION
**Pull request by indiosmo**
Useful when writing and debugging templates.

### CODE REVIEW

File: include/spore/codegen/codegen_app.hpp [hunk 1]
```diff
@@ -123,6 +123,18 @@ namespace spore::codegen
...
 
+                    if (options.dump_ast)
+                    {
+                        std::filesystem::path output_filename = fmt::format("{}.dump", input.stem().string());
+                        std::filesystem::path output_directory = std::filesystem::path(options.output) / "ast" / input.parent_path();
+                        std::string output = (output_directory / output_filename).string();
+
+                        if (!spore::codegen::write_file(output, json_data.dump(2)))
+                        {
+                            throw codegen_error(codegen_error_code::io, "failed to write output, file={}", output);
```
Comment (sporacid) [inline include/spore/codegen/codegen_app.hpp:134]: Should it fail if failing to dump the AST? I think we could log a warning in this case and it should be enough, what do you think?

**Review by sporacid — CHANGES_REQUESTED** [review]
Well done, it can be very useful to have access to the JSON structure to write templates. Now you'll be able to see that I'm missing some parsing in the AST! I've let a few comments, you can address those and I'll accept the PR. Thanks for your contribution!

**Comment by indiosmo** [comment]
`--dump-ast` is now a path, the dump is only generated if the option is set.

It no longer fails, but rather warns on error.

The extension has been changed to .json.

**Comment by sporacid** [comment]
> `--dump-ast` is now a path, the dump is only generated if the option is set.
> 
> It no longer fails, but rather warns on error.
> 
> The extension has been changed to .json.

I've updated the PR regex, you can rebase your branch and it should pass the PR check.

**Comment by indiosmo** [comment]
I'm not particularly familiar with the github pull request workflow, I used a rebase button and it looks like we're good?

Please note that I've made some changes since your last comment to support the option via the cmake integration.

**Comment by sporacid** [comment]
> I'm not particularly familiar with the github pull request workflow, I used a rebase button and it looks like we're good?
> 
> Please note that I've made some changes since your last comment to support the option via the cmake integration.

Me neither honestly, I'm learning on the go!

**Review by sporacid — CHANGES_REQUESTED** [review]

**Comment by indiosmo** [comment]
Should be good now.

**Pull request merged**